src/lightning/data/streaming/dataset.py

```diff
@@ -141,10 +170,26 @@ def __iter__(self) -> "StreamingDataset":
             self.worker_chunks.append(chunk_index)
             self.worker_intervals.append(chunk_interval)
 
-        self.current_indexes = []
-        self.chunk_index = 0
-        self.index = 0
+        # Handle restart
+        if self._state_dict:
+            state = self._state_dict[str(self.cache.rank)]
+
+            # re-generate indexes
+            interval = self.worker_intervals[self.chunk_index]
+            current_indexes = np.arange(interval[0], interval[1])
+            current_indexes = self.shuffler(current_indexes, self.current_epoch, self.chunk_index)
+            self.current_indexes = current_indexes[state["index"] :]
+
+            # Bump the chunk_index
+            self.chunk_index += 1
```

Comment:
Why +1 the index? We're reloading it in the line above. If the chunk wasn't complete, we would now miss the remainder?

Reply:
```python
  state = self._state_dict[str(self.cache.rank)]

  # re-generate indexes
  interval = self.worker_intervals[self.chunk_index]
  current_indexes = np.arange(interval[0], interval[1])
  current_indexes = self.shuffler(current_indexes, self.current_epoch, self.chunk_index)
  self.current_indexes = current_indexes[state["index"] :]

  # Bump the chunk_index
  self.chunk_index += 1
```

No, it won't. The `chunk_index` is bumped only once the  `current_indexes` are re-created.

I will clean this up in another PR.